### PR TITLE
[BugFix] Fix some bug of executing alter operation for tables separating primary key and sort key

### PR DESCRIPTION
**Pull request by sevev**
This pr fix two bugs:
1. Adding/dropping columns for tables with separate primary and sort keys could cause short index errors
2. Modifying the position of a non-sorted table could change the sorted key column

The reason of the bugs is we do not check if the origin tablet has sort key or not and use primary key to calculate short key column num. But the sort key column index still be assign to new created tablet.

## What type of PR is this:
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.1
  - [x] 3.0
  - [x] 2.5
  - [ ] 2.4

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 11 / 30 (36.67%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeHandler.java|11|30|36.67%|[1095, 1096, 1097, 1098, 1099, 1100, 1101, 1102, 1104, 1105, 1106, 1110, 1111, 1113, 1114, 1115, 1116, 1117, 1118]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 11 / 30 (36.67%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeHandler.java|11|30|36.67%|[1095, 1096, 1097, 1098, 1099, 1100, 1101, 1102, 1104, 1105, 1106, 1110, 1111, 1113, 1114, 1115, 1116, 1117, 1118]|

**Review by decster — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/alter/SchemaChangeHandler.java [hunk 1]
```diff
@@ -1088,12 +1088,41 @@ private AlterJobV2 createJob(long dbId, OlapTable olapTable, Map<Long, LinkedLis
...
-            LOG.debug("schema change[{}-{}-{}] check pass.", dbId, tableId, alterIndexId);
+            List<Integer> sortKeyIdxes = new ArrayList<>();
+            if (KeysType.PRIMARY_KEYS == olapTable.getKeysType()) {
+                MaterializedIndexMeta index = olapTable.getIndexMetaByIndexId(alterIndexId);
+                if (index.getSortKeyIdxes() != null) {
+                    List<Integer> originSortKeyIdxes = index.getSortKeyIdxes();
+                    for (Integer colIdx : originSortKeyIdxes) {
+                        String columnName = index.getSchema().get(colIdx).getName();
+                        Optional<Column> oneCol = 
+                                alterSchema.stream().filter(c -> c.getName().equalsIgnoreCase(columnName)).findFirst();
```
Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/alter/SchemaChangeHandler.java:1099]: If I have two columns in sort key columns, after dropping two columns?
Does the sort key to be primary key or throw an error?

Comment (sevev) [reply fe/fe-core/src/main/java/com/starrocks/alter/SchemaChangeHandler.java:1099]: > If I have two columns in sort key columns, after dropping two columns? Does the sort key to be primary key or throw an error?

We don't allow add/modify/drop sort key column so far, and these operations are reject in previous check.

**Review by Astralidea — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.1

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-2.5

**Pull request merged**